runtime/vm/jfr.cpp

```diff
@@ -1152,6 +1173,168 @@ jfrDump(J9VMThread *currentThread, BOOLEAN finalWrite)
 	flushAllThreadBuffers(currentThread, finalWrite);
 	writeOutGlobalBuffer(currentThread, finalWrite);
 }
+
+static UDATA
+jfrTypeIDHashFn(void *key, void *userData)
+{
+	const J9JFRTypeID *entry = (const J9JFRTypeID *)key;
+
+	return (UDATA)entry->clazz;
+}
+
+static UDATA
+jfrTypeIDHashEqualFn(void *tableNode, void *queryNode, void *userData)
+{
+	const J9JFRTypeID *tableEntry = (const J9JFRTypeID *)tableNode;
+	const J9JFRTypeID *queryEntry = (const J9JFRTypeID *)queryNode;
+
+	return tableEntry->clazz == queryEntry->clazz;
+}
+
+UDATA
+initializeJFRIDs(J9JavaVM *vm)
+{
+	UDATA result = 0;
+
+	if (omrthread_monitor_init_with_name(&vm->jfrState.typeIDMonitor, 0, "JFR Type ID monitor")) {
+		result = 1;
+		goto done;
+	}
+
+	vm->jfrState.typeIDcount = RESERVED_CLASS_ID_LIMIT;
+
+done:
+	return result;
+}
+
+void
+shutdownJFRIDs(J9JavaVM *vm)
+{
+	if (NULL != vm->jfrState.typeIDMonitor) {
+		omrthread_monitor_destroy(vm->jfrState.typeIDMonitor);
+		vm->jfrState.typeIDMonitor = NULL;
+	}
+}
+
+/**
+ * Known event types are only ever looked up by name and are always
+ * JVM event types so they can be below the RESERVED_CLASS_ID_LIMIT. The
+ * JVM determines the IDs for these types. They will always be loaded
+ * by the boot loader.
+ *
+ * TODO this function will be expanded in the future to provide general
+ * support for known JVM event types. Currently, the types supported are
+ * the ones needed to bootstrap JFR JCL initialization.
+ */
+jlong
+getKnownJFREventType(const J9UTF8 *className)
+{
+	jlong result = INVALID_TYPE_ID;
+	if (J9UTF8_LITERAL_EQUALS_UTF8(className, "boolean")) {
+		result = BOOLEAN_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "byte")) {
+		result = BYTE_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "char")) {
+		result = CHAR_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "short")) {
+		result = SHORT_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "int")) {
+		result = INT_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "float")) {
+		result = FLOAT_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "long")) {
+		result = LONG_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "double")) {
+		result = DOUBLE_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "jdk/types/StackTrace")) {
+		result = STACKTRACE_TYPE_ID;
+	}
+
+	return result;
+}
+
+jlong
+getTypeIDUTF8(J9VMThread *currentThread, const J9UTF8 *className)
```

Comment:
Do these functions need to acquire the mutex, or are they expecting it? I think this needs to be documented. I believe I saw a native caller above that did not acquire any mutex.

Reply:
Callers dont need a mutex. The mutex is acquired in `getTypeID`